File: EmojibleIOS/EmojibleIOS/DataSources/TeacherClassDataSource.swift
//
//  TeacherClassDataSource.swift
//  EmojibleIOS
//
//  Created by Furkan Yakal on 16.12.2020.
//

import Foundation
import Firebase

class TeacherClassDataSource{
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    var classroom: [ClassModel] = []
    var classroomDataSourceIndices: [String] = []
    
    var snapshotListenerForClassTutorials:ListenerRegistration?
    var classTutorials:[CodeModel] = []
    
    func startObservingClass(){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let classesCollectionRef = database.collection("Classes").whereField("teacherId", isEqualTo: currentUser.uid)
        snapshotListener = classesCollectionRef.addSnapshotListener{ [unowned self] querySnapshot, error in
            guard let documents = querySnapshot?.documents else { return }
            classroom = documents.map{ClassModel(dictionary: $0.data())}
            classroomDataSourceIndices = documents.map{$0.documentID}
            notificationCenter.post(name: .teacherClassChanged, object: nil, userInfo:["teacherClassChanged":classroom])
        }
    }
    
    func stopObservingClass(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()
    }
    
    func removeClass(classroom: ClassModel){
        var index: Int?
        for (i, document) in self.classroom.enumerated(){
            if document == classroom{
                index = i
                break
            }
        }
        UserDataSource.getInstance().resetUserClassIdForAllUsers(classId: classroom.id)
        database.collection("Classes").document(self.classroomDataSourceIndices[index!]).delete()
    }
    
    func writeClass(className:String, classPassword:String){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let uid = currentUser.uid
        let doc = self.database.collection("Classes").document()
        let newClassModel = ClassModel(id: String(doc.documentID.prefix(5)), teacherId:uid, className: className, password: classPassword)
        doc.setData(newClassModel.dictionary)
    }
    
    func checkClassOccurrence(name: String) -> Bool{
        for document in self.classroom{
            if document.className == name{
                return true
            }
        }
        return false
    }
    
    func addStudent(classId: String, classPassword:String, completion: @escaping (String?) -> Void){
        database.collection("Classes").whereField("id", isEqualTo: classId).whereField("password", isEqualTo: classPassword).getDocuments() { (querySnapshot, err) in
            if let err = err {
                completion("Error getting documents: \(err)")
            }
            else{
                guard let docs = querySnapshot?.documents else {
                    completion("Error getting documents: \(err)")
                    return
                }
                if docs.count == 0{
                    completion("No class with this ClassId and Password")
                }else {
                    UserDataSource.getInstance().editUserData(newData: ["classId" : classId])
                    completion(nil)
                }
                
            }
        }
    }
    
    func shareTutorial(classroom:ClassModel, code: CodeModel){
        var index: Int?
        for (i, document) in self.classroom.enumerated(){
            if document == classroom{
                index = i
                break
            }
        }
        self.database.collection("Classes").document(self.classroomDataSourceIndices[index!]).collection("ClassTutorials").addDocument(data:code.dictionary)
    }
    
    func startObservingClassTutorials(){
        UserDataSource.getInstance().getCurrentUserInfo(){ userModel in
            guard let userModel = userModel as? StudentModel else {return}
            guard let classId = userModel.classId else {return}
            self.database.collection("Classes").whereField("id", isEqualTo: classId).getDocuments() { (querySnapshot, err) in
                if let err = err {
                    print("Error getting documents: \(err)")
                }
                else{
                    guard let dict = querySnapshot?.documents[0].data() else {return}
                    let teacherId = dict["teacherId"] as! String
                    let classTutorialRef = self.database.collection("Users").document(teacherId).collection("Tutorials")
                    self.snapshotListenerForClassTutorials = classTutorialRef.addSnapshotListener{ [unowned self] docSnapshot, error in
                        guard let documents = docSnapshot?.documents else { return }
                        self.classTutorials = documents.map{CodeModel(dictionary: $0.data())}
                        notificationCenter.post(name: .classTutorialChanged, object: nil, userInfo:["classTutorialChanged":self.classTutorials])
                    }
                }
            }
        }
    }
    
    func stopObservingClassTutorials(){
        guard let snapshotListenerForClassTutorials = snapshotListener else { return }
        snapshotListenerForClassTutorials.remove()
    }
    
    func getClassName(completion: @escaping (String?) -> Void) {
        UserDataSource.getInstance().getCurrentUserInfo(){ userModel in
            guard let userModel = userModel as? StudentModel else {return}
            guard let classId = userModel.classId else {return}
            self.database.collection("Classes").whereField("id", isEqualTo: classId).getDocuments() {(querySnapshot, err) in
                if let dict = querySnapshot?.documents[0].data(){
                    let className = dict["className"] as? String
                    completion(className)
                }else {
                    completion(nil)
                }
            }
        }
    }
    
    private init(){}
    private static let instance = TeacherClassDataSource()
    public static func getInstance() -> TeacherClassDataSource{
        return .instance
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/TeacherClassDataSource.kt
package com.dji.emojibleandroid.dataSources

import android.os.Build
import androidx.annotation.RequiresApi
import com.dji.emojibleandroid.models.ClassModel
import com.dji.emojibleandroid.models.CodeModel
import com.dji.emojibleandroid.models.StudentModel
import com.dji.emojibleandroid.services.AuthenticationManager
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import java.lang.Exception

class TeacherClassDataSource private constructor(){
    val database = Firebase.firestore
    var snapshotListener: ListenerRegistration? = null
    val notificationCenter = NotificationCenter.instance
    var classroom = listOf<ClassModel>()
    var classRoomDataSourceIndices = listOf<String>()
    var snapshotListenerForClassTutorials: ListenerRegistration? = null
    var classTutorials = listOf<CodeModel>()

    fun startObservingClass() {
        val currentUser = AuthenticationManager.instance.currentUser ?: return
        val classesCollectionRef =
            database.collection("Classes").whereEqualTo("teacherId", currentUser.uid)
        snapshotListener = classesCollectionRef.addSnapshotListener { querySnapshot, error ->
            val documents = querySnapshot?.documents ?: return@addSnapshotListener
            classroom = documents.map { ClassModel(it.data as HashMap<String, Any>) }
            classRoomDataSourceIndices = documents.map { it.id }
            notificationCenter.post(
                Changes.teacherClassChanged,
                null,
                hashMapOf("teacherClassChanged" to classroom)
            )
        }
    }

    fun stopObservingClass() {
        val snapshotListener = snapshotListener ?: return
        snapshotListener.remove()
    }

    fun removeClass(classroom: ClassModel) {
        val cModel = this.classroom.find { it.equals(classroom) }
        val index = this.classroom.indexOf(cModel)
        UserDataSource.instance.resetUserClassIdForAllUsers(classroom.id)
        if (index != -1)
            database.collection("Classes").document(classRoomDataSourceIndices[index]).delete()
    }

    fun writeClass(className: String, classPassword: String) {
        val currentUser = AuthenticationManager.instance.currentUser ?: return
        val uid = currentUser.uid
        val doc = database.collection("Classes").document()
        val newClassModel = ClassModel(doc.id.substring(0..4), uid, className, classPassword)
        doc.set(newClassModel.dictionary)
    }

    fun checkClassOccurrence(name: String): Boolean {
        return classroom.any { it.className == name }
    }

    fun addStudent(classId: String, classPassword: String, completion: (String?) -> Unit) {
        database.collection("Classes").whereEqualTo("id", classId)
            .whereEqualTo("password", classPassword).get().addOnCompleteListener {
                if (it.isCanceled) {
                    completion("Error getting documents ${it.exception?.message}")
                } else if (it.isSuccessful) {
                    val docs = it.result?.documents
                    if (docs == null) {
                        completion("Error getting documents ${it.exception?.message}")
                        return@addOnCompleteListener
                    }
                    if (docs.isEmpty()) {
                        completion("No class with this id and password")
                    } else {
                        UserDataSource.instance.editUserData(hashMapOf("classId" to classId))
                        completion(null)
                    }
                }
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun startObservingClassTutorials() {
        UserDataSource.instance.getCurrentUserInfo { userModel ->
            val userModel = userModel as? StudentModel ?: return@getCurrentUserInfo
            val classId = userModel.classId ?: return@getCurrentUserInfo
            database.collection("Classes").whereEqualTo("id", classId).get().addOnCompleteListener { taskQuerySnapshot ->
                if (taskQuerySnapshot.isCanceled) {
                    throw Exception("Error getting documents ${taskQuerySnapshot.exception?.message}")
                } else if (taskQuerySnapshot.isSuccessful) {
                    val dict = taskQuerySnapshot.result?.documents?.get(0)?.data ?: return@addOnCompleteListener
                    val teacherId = dict["teacherId"] as String
                    val classTutorialRef = database.collection("Users").document(teacherId).collection("Tutorials")
                    snapshotListenerForClassTutorials = classTutorialRef.addSnapshotListener { docSnapshot, error ->
                        val documents = docSnapshot?.documents ?: return@addSnapshotListener
                        classTutorials = documents.map { CodeModel(it.data as HashMap<String, Any>) }
                        notificationCenter.post(Changes.classTutorialChanged, null, hashMapOf("classTutorialChanged" to classTutorials))
                    }
                }
            }
        }
    }

    fun stopObservingClassTutorials() {
        val snapshotListenerForClassTutorials = snapshotListenerForClassTutorials ?: return
        snapshotListenerForClassTutorials.remove()
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun getClassName(completion: (String?) -> Unit) {
        UserDataSource.instance.getCurrentUserInfo { userModel ->
            val userModel = userModel as? StudentModel ?: return@getCurrentUserInfo
            val classId = userModel.classId ?: return@getCurrentUserInfo
            database.collection("Classes").whereEqualTo("id", classId).get().addOnCompleteListener {
                val dict = it.result?.documents?.get(0)?.data
                if (dict == null) {
                    completion(null)
                } else {
                    val className = dict["className"] as? String
                    completion(className)
                }
            }
        }
    }

    private object HOLDER {
        val INSTANCE = TeacherClassDataSource()
    }

    companion object {
        val instance: TeacherClassDataSource by lazy { HOLDER.INSTANCE }
    }
}